lib/open_food_network/scope_variants_for_search.rb

```diff
@@ -34,7 +34,10 @@ def search_params
     def query_scope
       Spree::Variant.where(is_master: false).
         includes(option_values: :option_type).
-        ransack(search_params.merge(m: 'or')).result
+        ransack(search_params.merge(m: 'or')).
+        result.
+        includes(:product).
+        order("products_spree_variants.name, display_name, display_as, products_spree_variants.variant_unit_name")
```

Comment:
Where does that table name come from? Isn't it `spree_products`?

There are failing specs related to this.

Reply:
@mkllnk 
I don't see any solution here, I'm a bit stuck actually. 
`products_spree_variants ` seems to be a generated table name from ``.
See the generated SQL here: 
```sql
SELECT DISTINCT spree_variants.*,
                "spree_variants"."id" AS t0_r0,
                "spree_option_values"."id" AS t1_r0,
                "spree_option_values"."position" AS t1_r1,
                "spree_option_values"."name" AS t1_r2,
                "spree_option_values"."presentation" AS t1_r3,
                "spree_option_values"."option_type_id" AS t1_r4,
                "spree_option_values"."created_at" AS t1_r5,
                "spree_option_values"."updated_at" AS t1_r6,
                "spree_option_types"."id" AS t2_r0,
                "spree_option_types"."name" AS t2_r1,
                "spree_option_types"."presentation" AS t2_r2,
                "spree_option_types"."created_at" AS t2_r3,
                "spree_option_types"."updated_at" AS t2_r4,
                "spree_option_types"."position" AS t2_r5,
                "products_spree_variants"."id" AS t3_r0,
                "products_spree_variants"."name" AS t3_r1,
                "products_spree_variants"."description" AS t3_r2,
                "products_spree_variants"."available_on" AS t3_r3,
                "products_spree_variants"."deleted_at" AS t3_r4,
                "products_spree_variants"."permalink" AS t3_r5,
                "products_spree_variants"."meta_description" AS t3_r6,
                "products_spree_variants"."meta_keywords" AS t3_r7,
                "products_spree_variants"."tax_category_id" AS t3_r8,
                "products_spree_variants"."shipping_category_id" AS t3_r9,
                "products_spree_variants"."created_at" AS t3_r10,
                "products_spree_variants"."updated_at" AS t3_r11,
                "products_spree_variants"."supplier_id" AS t3_r12,
                "products_spree_variants"."group_buy" AS t3_r13,
                "products_spree_variants"."group_buy_unit_size" AS t3_r14,
                "products_spree_variants"."variant_unit" AS t3_r15,
                "products_spree_variants"."variant_unit_scale" AS t3_r16,
                "products_spree_variants"."variant_unit_name" AS t3_r17,
                "products_spree_variants"."notes" AS t3_r18,
                "products_spree_variants"."primary_taxon_id" AS t3_r19,
                "products_spree_variants"."inherits_properties" AS t3_r20
FROM "spree_variants"
INNER JOIN "exchange_variants" ON "exchange_variants"."variant_id" = "spree_variants"."id"
INNER JOIN "exchanges" ON "exchanges"."id" = "exchange_variants"."exchange_id"
INNER JOIN "order_cycles" ON "order_cycles"."id" = "exchanges"."order_cycle_id"
LEFT OUTER JOIN "spree_products" ON "spree_products"."id" = "spree_variants"."product_id"
LEFT OUTER JOIN "spree_option_values_variants" ON "spree_option_values_variants"."variant_id" = "spree_variants"."id"
LEFT OUTER JOIN "spree_option_values" ON "spree_option_values"."id" = "spree_option_values_variants"."option_value_id"
LEFT OUTER JOIN "spree_option_types" ON "spree_option_types"."id" = "spree_option_values"."option_type_id"
LEFT OUTER JOIN "spree_products" "products_spree_variants" ON "products_spree_variants"."id" = "spree_variants"."product_id"
WHERE "spree_variants"."deleted_at" IS NULL
  AND "spree_variants"."is_master" = $1
  AND ("spree_products"."name" ILIKE '%brea%'
       OR "spree_variants"."sku" ILIKE '%brea%')
  AND "exchanges"."incoming" = $2
  AND (order_cycles.id = '1')
  AND "spree_variants"."id" IN
    (SELECT "exchange_variants"."variant_id"
     FROM "exchange_variants"
     INNER JOIN "exchanges" ON "exchanges"."id" = "exchange_variants"."exchange_id"
     WHERE (exchanges.incoming = FALSE
            AND exchanges.receiver_id = 3))
ORDER BY products_spree_variants.name,
         display_name,
         display_as,
         products_spree_variants.variant_unit_name
```



If I used `spree_products` instead (which seems to be logical) I get a SQL error: 
`ORDER BY expressions must appear in select list`
The generated SQL is here: 
```sql
SELECT DISTINCT spree_variants.*
FROM "spree_variants"
INNER JOIN "exchange_variants" ON "exchange_variants"."variant_id" = "spree_variants"."id"
INNER JOIN "exchanges" ON "exchanges"."id" = "exchange_variants"."exchange_id"
INNER JOIN "order_cycles" ON "order_cycles"."id" = "exchanges"."order_cycle_id"
LEFT OUTER JOIN "spree_products" ON "spree_products"."id" = "spree_variants"."product_id"
WHERE "spree_variants"."deleted_at" IS NULL
  AND "spree_variants"."is_master" = $1
  AND ("spree_products"."name" ILIKE '%bread%'
       OR "spree_variants"."sku" ILIKE '%bread%')
  AND "exchanges"."incoming" = $2
  AND (order_cycles.id = '1')
  AND "spree_variants"."id" IN
    (SELECT "exchange_variants"."variant_id"
     FROM "exchange_variants"
     INNER JOIN "exchanges" ON "exchanges"."id" = "exchange_variants"."exchange_id"
     WHERE (exchanges.incoming = FALSE
            AND exchanges.receiver_id = 3))
ORDER BY spree_products.name,
         display_name,
         display_as,
         spree_products.variant_unit_name
```

I'd invoke some help here ;) Thanks a lot.

Reply:
Ouch, that's complicated. So in this case Rails generates the table name `products_spree_variants` selects attributes on that one. I don't know the solution. I need to play with it myself.

Reply:
Okay, I found two solutions. First, it occurred to me that Rails has a way of telling that you want to reference the joined table and it looked like this:
```diff
--- a/lib/open_food_network/scope_variants_for_search.rb
+++ b/lib/open_food_network/scope_variants_for_search.rb
@@@ -37,7 -37,7 +37,8 @@@ module OpenFoodNetwor
          ransack(search_params.merge(m: 'or')).
          result.
          includes(:product).
--        order("products_spree_variants.name, display_name, display_as, products_spree_variants.variant_unit_name")
++        references(:product).
++        order("spree_products.name, display_name, display_as, spree_products.variant_unit_name")
      end
  
      def distributor
```

That's great but the generated query is ugly and I found a better way of telling Rails how to sort and it can optimise under the hood however it likes. I added a commit for that. :crossed_fingers: I hope that all specs pass now.

Reply:
Okay, this didn't work. We have more failing specs now. The problem is that some code parts are changing the select statement, I think. Maybe we have to include the order statement somewhere else.

Reply:
Based on what I've done in https://github.com/openfoodfoundation/openfoodnetwork/pull/8615, I made another try with commit: a33ebbdc8
```ruby
[...]
        result.
        order("spree_products.name, display_name, display_as, spree_products.variant_unit_name").
        includes(:product).
        joins(:product)
```
The key here is to `include` and `joins` _after_ the `order` in order to `select` `spree_products.*` and therefore be able to `order by`  `spree_products.name` and `spree_products.variant_unit_name`

Reply:
@jibees Could you paste the new SQL request?

Reply:
@achauve 

```sql
SELECT DISTINCT spree_variants.*,
                "spree_variants"."id" AS t0_r0,
                "spree_option_values"."id" AS t1_r0,
                "spree_option_values"."position" AS t1_r1,
                "spree_option_values"."name" AS t1_r2,
                "spree_option_values"."presentation" AS t1_r3,
                "spree_option_values"."option_type_id" AS t1_r4,
                "spree_option_values"."created_at" AS t1_r5,
                "spree_option_values"."updated_at" AS t1_r6,
                "spree_option_types"."id" AS t2_r0,
                "spree_option_types"."name" AS t2_r1,
                "spree_option_types"."presentation" AS t2_r2,
                "spree_option_types"."created_at" AS t2_r3,
                "spree_option_types"."updated_at" AS t2_r4,
                "spree_option_types"."position" AS t2_r5,
                "spree_products"."id" AS t3_r0,
                "spree_products"."name" AS t3_r1,
                "spree_products"."description" AS t3_r2,
                "spree_products"."available_on" AS t3_r3,
                "spree_products"."deleted_at" AS t3_r4,
                "spree_products"."permalink" AS t3_r5,
                "spree_products"."meta_description" AS t3_r6,
                "spree_products"."meta_keywords" AS t3_r7,
                "spree_products"."tax_category_id" AS t3_r8,
                "spree_products"."shipping_category_id" AS t3_r9,
                "spree_products"."created_at" AS t3_r10,
                "spree_products"."updated_at" AS t3_r11,
                "spree_products"."supplier_id" AS t3_r12,
                "spree_products"."group_buy" AS t3_r13,
                "spree_products"."group_buy_unit_size" AS t3_r14,
                "spree_products"."variant_unit" AS t3_r15,
                "spree_products"."variant_unit_scale" AS t3_r16,
                "spree_products"."variant_unit_name" AS t3_r17,
                "spree_products"."notes" AS t3_r18,
                "spree_products"."primary_taxon_id" AS t3_r19,
                "spree_products"."inherits_properties" AS t3_r20
FROM "spree_variants"
INNER JOIN "spree_products" ON "spree_products"."id" = "spree_variants"."product_id"
INNER JOIN "exchange_variants" ON "exchange_variants"."variant_id" = "spree_variants"."id"
INNER JOIN "exchanges" ON "exchanges"."id" = "exchange_variants"."exchange_id"
INNER JOIN "order_cycles" ON "order_cycles"."id" = "exchanges"."order_cycle_id"
LEFT OUTER JOIN "spree_option_values_variants" ON "spree_option_values_variants"."variant_id" = "spree_variants"."id"
LEFT OUTER JOIN "spree_option_values" ON "spree_option_values"."id" = "spree_option_values_variants"."option_value_id"
LEFT OUTER JOIN "spree_option_types" ON "spree_option_types"."id" = "spree_option_values"."option_type_id"
WHERE "spree_variants"."deleted_at" IS NULL
  AND "spree_variants"."is_master" = $1
  AND ("spree_products"."name" ILIKE '%bread%'
       OR "spree_variants"."sku" ILIKE '%bread%')
  AND "exchanges"."incoming" = $2
  AND (order_cycles.id = '1')
  AND "spree_variants"."id" IN
    (SELECT "exchange_variants"."variant_id"
     FROM "exchange_variants"
     INNER JOIN "exchanges" ON "exchanges"."id" = "exchange_variants"."exchange_id"
     WHERE (exchanges.incoming = FALSE
            AND exchanges.receiver_id = 3))
ORDER BY spree_products.name,
         display_name,
         display_as,
         spree_products.variant_unit_name
```

Reply:
I guess the order by `display_name` and `display_as` is useless...

Reply:
Thanks! your query seems a lot cleaner than before! However I'm still not sure why. There are lots of nested rails scopes. Is it usefull only to override previous scopes? And in your case don't "joins" and "includes" do the same thing (= inner join to products table to include products in result set)?